engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java

```diff
@@ -130,7 +122,15 @@ private AbstractHSQuery(ExtendedSearchIntegrator extendedIntegrator, IndexedType
 				throw LOG.targetedEntityTypesNotIndexed( StringHelper.join( targetedEntities, "," ) );
 			}
 		}
-		this.customTypeMetadata = customTypeMetadata == null ? Collections.emptyList() : new ArrayList<>( customTypeMetadata );
+		if ( customTypeMetadata == null ) {
+			this.customTypeMetadata = IndexedTypeMaps.empty();
+		}
+		else {
+			this.customTypeMetadata = IndexedTypeMaps.hashMap();
+			for ( Map.Entry<IndexedTypeIdentifier, CustomTypeMetadata> entry : customTypeMetadata.entrySet() ) {
```

Comment:
It was mainly because `customTypeMetadata` is basically "user" input, so I can't trust it not to change. As you said, we want to make the map immutable _eventually_, but we're not quite there yet :)
Anyway, ok to remove the copy, it's not that critical.